mbf_abstract_nav/src/move_base_action.cpp

```diff
@@ -454,45 +449,41 @@ void MoveBaseAction::actionRecoveryDone(
   last_oscillation_reset_ = ros::Time::now();
 
   const mbf_msgs::RecoveryResult& recovery_result = *result_ptr;
-  mbf_msgs::MoveBaseResult move_base_result;
-
-  // copy result from recovery action
-  fillMoveBaseResult(recovery_result, move_base_result);
 
   switch (state.state_)
   {
     case actionlib::SimpleClientGoalState::REJECTED:
     case actionlib::SimpleClientGoalState::ABORTED:
-      action_state_ = FAILED;
-
-      ROS_DEBUG_STREAM_NAMED("move_base", "The recovery behavior \""
-          << *current_recovery_behavior_ << "\" has failed. ");
+    case actionlib::SimpleClientGoalState::RECALLED:
+    case actionlib::SimpleClientGoalState::PREEMPTED:
+      ROS_WARN_STREAM_NAMED("move_base", "The recovery behavior \"" << *current_recovery_behavior_ << "\" has been "
+                                                                    << state.toString());
       ROS_DEBUG_STREAM("Recovery behavior message: " << recovery_result.message
-                                    << ", outcome: " << recovery_result.outcome);
+                                                     << ", outcome: " << recovery_result.outcome);
 
-      current_recovery_behavior_++; // use next behavior;
-      if (current_recovery_behavior_ == recovery_behaviors_.end())
+      if (action_state_ == CANCELED)
       {
-        ROS_DEBUG_STREAM_NAMED("move_base",
-                               "All recovery behaviors failed. Abort recovering and abort the move_base action");
-        goal_handle_.setAborted(move_base_result, "All recovery behaviors failed.");
+        // move_base preempted while executing a recovery; fill result and report canceled to the client
+        fillMoveBaseResult(recovery_result, move_base_result_);
+        ROS_INFO_STREAM_NAMED("move_base", "move_base preempted while executing a recovery behavior");
+        goal_handle_.setCanceled(move_base_result_, state.getText());
+        break;
       }
-      else
+      if (current_recovery_behavior_ == recovery_behaviors_.end())
       {
-        recovery_goal_.behavior = *current_recovery_behavior_;
-
-        ROS_INFO_STREAM_NAMED("move_base", "Run the next recovery behavior \""
-            << *current_recovery_behavior_ << "\".");
-        action_client_recovery_.sendGoal(
-            recovery_goal_,
-            boost::bind(&MoveBaseAction::actionRecoveryDone, this, _1, _2)
-        );
+        // we run out of recovery behaviors; concede defeat
+        action_state_ = FAILED;
+        ROS_DEBUG_STREAM_NAMED("move_base", "All recovery behaviors failed. Aborting move_base action");
+        goal_handle_.setAborted(move_base_result_, "All recovery behaviors failed.");
+        break;
       }
-      break;
+      // recovery failed but we still have behaviors; we cannot call the next here (simple action client
+      // will complain about transition to ACTIVE from DONE), so we try replanning again (wrong, but works)
     case actionlib::SimpleClientGoalState::SUCCEEDED:
```

Comment:
```suggestion
      // recovery failed but we still have behaviors; we cannot call the next here
      // (simple action client will complain about transition to ACTIVE from DONE)
      // so we try replanning again (wrong, but works)
      [[fallthrough]];
    case actionlib::SimpleClientGoalState::SUCCEEDED:
```
S: add [`[[fallthrough]]`](https://en.cppreference.com/w/cpp/language/attributes/fallthrough) attribute to make it clear what the comment is referring to.
I was a bit confused at first until I noticed there was no `break` statement 😅
I think it requires C++17 though, so if you don't want to change `CMakeLists.txt` then maybe add it commented out

S: better text line breaks